analyzer/checker/src/engine.rs

```diff
@@ -36,8 +37,38 @@ impl<'a> Engine<'a> {
         SourceObjectId(id)
     }
 
-    /// Attaches an environment to an origin.
-    pub fn attach(&mut self, id: SourceObjectId, env: Environment) {
+    /// Attaches an environment to an origin if the origin does not already have an attached environment.
+    pub fn attach(&mut self, id: SourceObjectId, env: Environment) -> Result<(), String> {
+        if self.origins[id.0].1.is_some() {
+            return Err(format!(
+                "Could not attach environment to {}: an environment is already attached",
+                id.0
+            ));
+        }
         self.origins[id.0].1 = Some(env);
+        Ok(())
+    }
```

Comment:
```suggestion
    pub fn attach(&mut self, id: SourceObjectId, env: Environment) {
        debug_assert!(self.origins[id.0].1.is_none(), "Could not attach environment to a source that is already attached");
        self.origins[id.0].1.replace(env);
    }
```
If `attach` is called multiple times for the same object id, it is just a misuse of the API that would not benefit from a `Result`.